File: backend/src/controllers/v1/agents.py
from typing import Any, Dict, Optional
from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, Field, ValidationError
from ...logging import get_logger
from ...dao import agents as AgentsDao
from ...models.agent_model import Agent
from ...models.prompt_base import PromptSection

router = APIRouter(prefix="/agents", tags=["agents"])
logger = get_logger("controllers.agents")


class AgentNodePropertiesRequest(BaseModel):
    """Request model for creating or updating an agent"""
    name: str = Field(..., description="Name of the agent, e.g. 'WeatherBot'")
    description: str = Field(..., description="Description of the agent's purpose and capabilities")
    prompt_template: str = Field(..., description="The prompt template with placeholders for variables")
    variables: Dict[str, Any] = Field(default_factory=dict, description="Key-value pairs for variables used in the prompt template")
    version: float = Field(..., description="Version number for the prompt template")
    sections: Optional[list[PromptSection]] = Field(default_factory=list, description="List of additional prompt sections")


class PaginatedAgentsResponse(BaseModel):
    """Response model for paginated agents list"""
    agents: list[Agent]
    total: int
    limit: int
    offset: int


@router.get("", response_model=PaginatedAgentsResponse)
def list_agents(
    limit: int = Query(default=50, ge=1, le=100, description="Maximum number of agents to return"),
    offset: int = Query(default=0, ge=0, description="Number of agents to skip")
):
    """
    List agents with pagination
    
    Args:
        limit: Maximum number of agents to return (1-100)
        offset: Number of agents to skip
    
    Returns:
        PaginatedAgentsResponse with agents list, total count, limit, and offset
    """
    logger.info(f"Listing agents with limit={limit}, offset={offset}")
    
    try:
        agents, total = AgentsDao.list_agents_paginated(limit, offset)
        return PaginatedAgentsResponse(
            agents=agents,
            total=total,
            limit=limit,
            offset=offset
        )
    except Exception as e:
        logger.error(f"Failed to list agents: {str(e)}")
        raise HTTPException(status_code=500, detail="Failed to list agents")


@router.post("", response_model=Agent, status_code=201)
def create_agent(
    request: AgentNodePropertiesRequest
):
    """
    Create a new agent
    
    Args:
        request: Agent properties
    
    Returns:
        Created Agent
    
    Raises:
        HTTPException 409: Agent with the same name already exists
        HTTPException 422: Invalid agent properties
    """
    logger.info(f"Creating agent: {request.name}")
    
    try:
        from ...models.agent_model import AgentNodeProperties
        props = AgentNodeProperties(**request.model_dump())
        agent = AgentsDao.create_agent(props)
        logger.info(f"Created agent with id={agent.id}")
        return agent
    except ValueError as e:
        if "already exists" in str(e):
            logger.warning(f"Duplicate agent name: {request.name}")
            raise HTTPException(status_code=409, detail=str(e))
        logger.error(f"Failed to create agent: {str(e)}")
        raise HTTPException(status_code=404, detail=str(e))
    except ValidationError as e:
        logger.error(f"Invalid agent properties: {str(e)}")
        raise HTTPException(status_code=422, detail=str(e))
    except Exception as e:
        logger.error(f"Unexpected error creating agent: {str(e)}")
        raise HTTPException(status_code=500, detail="Failed to create agent")


@router.get("/{agent_id}", response_model=Agent)
def get_agent(
    agent_id: str
):
    """
    Get an agent by ID
    
    Args:
        agent_id: Agent ID
    
    Returns:
        Agent
    
    Raises:
        HTTPException 404: Agent not found
    """
    logger.info(f"Getting agent with id={agent_id}")
    
    try:
        agent = AgentsDao.get_agent(agent_id)
        return agent
    except ValueError as e:
        logger.warning(f"Agent not found: {agent_id}")
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        logger.error(f"Failed to get agent: {str(e)}")
        raise HTTPException(status_code=500, detail="Failed to get agent")


@router.put("/{agent_id}", response_model=Agent)
def update_agent(
    agent_id: str,
    request: AgentNodePropertiesRequest
):
    """
    Update an agent by ID
    
    Args:
        agent_id: Agent ID
        request: Updated agent properties
    
    Returns:
        Updated Agent
    
    Raises:
        HTTPException 404: Agent not found
        HTTPException 409: Agent with the same name already exists
        HTTPException 422: Invalid agent properties
    """
    logger.info(f"Updating agent with id={agent_id}")
    
    try:
        from ...models.agent_model import AgentNodeProperties
        props = AgentNodeProperties(**request.model_dump())
        agent = AgentsDao.update_agent(agent_id, props)
        logger.info(f"Updated agent with id={agent_id}")
        return agent
    except ValueError as e:
        if "already exists" in str(e):
            logger.warning(f"Duplicate agent name: {request.name}")
            raise HTTPException(status_code=409, detail=str(e))
        logger.warning(f"Agent not found: {agent_id}")
        raise HTTPException(status_code=404, detail=str(e))
    except ValidationError as e:
        logger.error(f"Invalid agent properties: {str(e)}")
        raise HTTPException(status_code=422, detail=str(e))
    except Exception as e:
        logger.error(f"Unexpected error updating agent: {str(e)}")
        raise HTTPException(status_code=500, detail="Failed to update agent")


@router.delete("/{agent_id}", status_code=204)
def delete_agent(
    agent_id: str
):
    """
    Delete an agent by ID (hard delete)
    
    Args:
        agent_id: Agent ID
    
    Returns:
        None (204 No Content)
    
    Raises:
        HTTPException 404: Agent not found
    """
    logger.info(f"Deleting agent with id={agent_id}")
    
    try:
        # First check if agent exists
        AgentsDao.get_agent(agent_id)
        
        # Delete the agent
        success = AgentsDao.delete_agent(agent_id)
        if not success:
            raise ValueError(f"Failed to delete agent with id '{agent_id}'")
        
        logger.info(f"Deleted agent with id={agent_id}")
        return None
    except ValueError as e:
        logger.warning(f"Agent not found: {agent_id}")
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        logger.error(f"Failed to delete agent: {str(e)}")
        raise HTTPException(status_code=500, detail="Failed to delete agent")

File: backend/src/controllers/v1/agents.ts
import { Router, Request } from 'express';
import AgentDao from '../../lib/dao/agent.dao.js';
import MemoryDao from '../../lib/dao/memory.dao.js';
import { AgentProperties, CreateAgentDTO } from '../../lib/models/agent.js';
import { ZodBodyValidator, ZodIdValidator, ZodQueryValidator } from '../../middleware/zod.middleware.js';
import { PaginationQuerySchemaBase, PaginationQuery } from '../../lib/types/pagination.js';
import { BadRequestError, NotFoundError } from '../../lib/errors/http.errors.js';
import { AgentRuntime } from '../../lib/agents/agent-runtime.js';

export const router = Router();

// Create a new agent
router.post('/',
  ZodBodyValidator(AgentProperties),
  async (req, res) => {
  req.logger.info('Loading Agent Manager');
  const agentManager = req.app.agents;
  
  try {
    req.logger.info('Creating new agent');
    
    const agentData: CreateAgentDTO = req.body;
    const agent = await AgentDao.createAgent(agentData);

    req.logger.debug('Agent created successfully', { ...agent });
    req.logger.info('Registering agent with Agent Manager');
    
    const llmEngine = req.app.llm.getClient(agentData.engine);
    agentManager.registerAgent(
      AgentRuntime.fromDatabase(agent, llmEngine)
    );
    
    req.logger.info('Agent registered successfully');

    res.status(201).json(agent);
  } catch (error) {
    req.logger.error('Error creating agent:', error);
    res.status(500).json({ error: 'Failed to create agent' });
  }
});

// List all agents (latest version only) with pagination
router.get('/',
  ZodQueryValidator(PaginationQuerySchemaBase),
  async (req, res) => {
  const agentManager = req.app.agents;

  const { page, take } = res.locals.query as any;

  const paginationQuery: PaginationQuery = {
    page,
    take,
    skip: (page - 1) * take
  };

  const paginatedAgents = await AgentDao.listAgents(paginationQuery);
  
  res.json({
    data: paginatedAgents.data.map(agent => ({
      ...agent,
      is_active: agentManager.isActive(agent.agent_id)
    })),
    pagination: paginatedAgents.pagination
  });
});

async function getAgentById(req: Request) {
  const agentId = Array.isArray(req.params.id) ?
    parseInt(req.params.id[0], 10) :
    parseInt(req.params.id, 10);

  if (isNaN(agentId)) {
    throw new BadRequestError('Invalid agent ID');
  }

  const agent = await AgentDao.getAgent(agentId);
  if (!agent) {
    throw new NotFoundError('Agent not found');
  }

  return agent;
}

// List active agents (lightweight)
router.get('/active', async (req, res) => {
  const agentManager = req.app.agents;
  const activeAgents = agentManager.listActiveAgents().map(runtime => ({
    agent_id: runtime.id,
    name: runtime.name,
    description: runtime.description,
  }));

  res.json({ agents: activeAgents });
});

// Get a specific agent by ID (latest version)
router.get('/:id',
  ZodIdValidator('id'),
  async (req, res) => {
    const agent = await getAgentById(req);
    res.json(agent);
  }
);

// Get agent details including memories and tools
router.get('/:id/details',
  ZodIdValidator('id'),
  async (req, res) => {
    const agentManager = req.app.agents;
    const agent = await getAgentById(req);

    const memories = await MemoryDao.listMemories(agent.agent_id);

    const agentRuntime = agentManager.getAgent(agent.agent_id);

    const agentTools = await agentRuntime?.getTools();
    agentTools
      ?.map(tool => tool.getName())
      .reduce((acc, toolName) => {
        acc[toolName] = true;
        return acc;
      }, {} as Record<string, boolean>);

    res.json({
      ...agent,
      is_active: agentManager.isActive(agent.agent_id),
      memories,
      tools: agentTools
      ?.map(tool => tool.getName())
      .reduce((acc, toolName) => {
        acc[toolName] = { allowEdit: false, value: true };
        return acc;
      }, {} as Record<string, { allowEdit: boolean, value: boolean }>)
    });
  }
);

router.post('/:id/start',
  ZodIdValidator(),
  async (req, res) => {
    req.logger.info('Loading Agent Manager');
    const agentManager = req.app.agents;

    const agent = await getAgentById(req);

    if (agentManager.isActive(agent.agent_id)) {
      throw new NotFoundError('Agent not found');
    }

    agentManager.startAgent(agent.agent_id);

    req.logger.info('Agent started successfully');
    res.status(200).json({ message: 'Agent started' });
  }
);

router.post('/:id/stop',
  ZodIdValidator(),
  async (req, res) => {
    req.logger.info('Loading Agent Manager');
    const agentManager = req.app.agents;
    
    const agent = await getAgentById(req);
    
    req.logger.info('Stopping agent', { agent: agent.name });
    agentManager.stopAgent(agent.agent_id);
    
    req.logger.info('Agent stopped successfully');
    res.status(200).json({ message: 'Agent stopped' });
  }
);

// Create a new version of an agent, optionally with modifications
router.post('/:id/version',
  ZodIdValidator('id'),
  ZodBodyValidator(AgentProperties.partial()),
  async (req, res) => {
    const agent = await getAgentById(req);

    try {
      const overrides: Partial<CreateAgentDTO> = req.body;
      const newVersion = await AgentDao.createAgentVersion(agent.agent_id, overrides);
      res.status(201).json(newVersion);
    } catch (error) {
      req.logger.error('Error creating agent version:', error);
      res.status(500).json({ error: 'Failed to create agent version' });
    }
  }
);

// Update an agent
router.put('/:id',
  ZodIdValidator('id'),
  ZodBodyValidator(AgentProperties.partial()),
  async (req, res) => {
  try {
    const agentId = Array.isArray(req.params.id) ?
      parseInt(req.params.id[0], 10) :
      parseInt(req.params.id, 10);
      
    if (isNaN(agentId)) {
      res.status(400).json({ error: 'Invalid agent ID' });
      return;
    }

    const agentData: Partial<CreateAgentDTO> = req.body;
    const updatedAgent = await AgentDao.updateAgent(agentId, agentData);
    res.json(updatedAgent);
  } catch (error) {
    req.logger.error('Error updating agent:', error);
    if (error instanceof Error && error.message === 'Agent not found') {
      res.status(404).json({ error: 'Agent not found' });
      return;
    }
    res.status(500).json({ error: 'Failed to update agent' });
  }
});

// Delete an agent (all versions)
router.delete(
  '/:id',
  ZodIdValidator('id'),
  async (req, res) => {
  try {
    const agentId = Array.isArray(req.params.id) ?
      parseInt(req.params.id[0], 10) :
      parseInt(req.params.id, 10);

    if (isNaN(agentId)) {
      res.status(400).json({ error: 'Invalid agent ID' });
      return;
    }

    await AgentDao.deleteAgent(agentId);
    res.status(204).send();
  } catch (error) {
    console.error('Error deleting agent:', error);
    res.status(500).json({ error: 'Failed to delete agent' });
  }
});

// Delete a specific memory belonging to an agent
router.delete('/:id/memories/:nodeId',
  ZodIdValidator('id'),
  ZodIdValidator('nodeId'),
  async (req, res) => {
    const agent = await getAgentById(req);
    const nodeId = (req.params as any).nodeId as number;

    const deleted = await MemoryDao.deleteMemory(nodeId, agent.agent_id);
    if (!deleted) {
      throw new NotFoundError('Memory not found');
    }

    res.status(204).send();
  }
);

export default router;